src/osd/OSD.cc

```diff
@@ -4372,6 +4374,65 @@ int OSD::handle_pg_peering_evt(
   }
 }
 
+bool OSD::maybe_wait_for_max_pg(int primary, spg_t pgid)
+{
+  const auto max_pgs_per_osd =
+    (cct->_conf->get_val<int64_t>("mon_max_pg_per_osd") *
+     cct->_conf->get_val<double>("osd_max_pg_per_osd_hard_ratio"));
+  RWLock::RLocker l(pg_map_lock);
+  if (pg_map.size() < max_pgs_per_osd) {
+    return false;
+  }
+  if (primary == whoami) {
+    want_to_create_primary++;
+  } else {
+    want_to_create_replica.emplace(pgid.pgid, primary);
+  }
```

Comment:
I don't think this has anything to do with whether we're primary... we can also get a MPGNotify from a replica/stray telling us we are the primary and need to instantiate the PG.  Instead of passing in int primary, maybe pass in bool is_create, and set that to true from handle_pg_create, false otherwise?  (And rename the want_to_create_replica to want_to_instantiate or something.)

Reply:
will do.

Reply:
@liewegas i think it's relevant. `maybe_wait_for_max_pg()` is called by `OSDD::handle_pg_peering_evt()` , right before the pg is about to be created. the reason why a replica/stray sends us a MOSDPGNotify is that it gets a `MOSDPGQuery` from us. in that case, we are primary, and we created the pg alreay.

Reply:
We can create a local PG both if we're primary or if we're a replica, though. For example,
1. pg is on [0,1], and osd.0 is reweighted.  new mapping becomes [2,1].  osd.2 is primary and needs to create the pg.
2. pg is on [0,1], and pool size is increased from 2 -> 3, changing mapping to [0,1,2].  osd.2 is replica and needs to create the pg.

Reply:
ah, i see. in the first case, there is chance that a replica/stray sends a `MOSDPGNotify` to the new primary. and if it has not been updated with the latest osdmap from the monitor, the new primary will create a pg as required by `MOSDPGNotify`.